Remove GoldenEye TLB hack

loganmc10 opened the pull request:
Fixes https://github.com/mupen64plus/mupen64plus-core/issues/447

This will need to be tested on every dynarec/interpreter. I've tested the Pure/Cached Interpreter as well as the 64-bit dynarec. I tested the x86 new dynarec a while ago, but it probably should be tested again. ARM should also be tested ( @fzurita )

The only game that should be affected is GoldenEye, however I did make a change to how the Pure Interpreter works generally.

loganmc10 commented:
Just re-tested the x86 new dynarec and it still works without the hack

fzurita commented:
The game seems to start and run. Performance seems a lot worse though, at least in Android. Was this a performance hack?

loganmc10 commented:
Well it's possible when it was first implemented it was necessary for some reason, but now it is basically a performance hack.

GoldenEye is a strange game in that is will purposely trigger TLB exceptions (it will try to execute code that isn't TLB mapped, which triggers a TLB exception, which will then map the TLB entry, and then try to execute the code again).

The hack bypasses the TLB exceptions, and points to the address needed directly. So yes I would except that CPU usage would go up since all these TLB exceptions need to be processed.

It still runs fine on my laptop, the new dynarec implements its own TLB code so there is the chance that this affects the new dynarec more negativity than the other methods? I'm not really sure.

loganmc10 commented:
So potentially we could leave the hack in place for the new dynarec, since that is mostly targeted at mobile devices? thoughts @bsmiles32 @Gillou68310 ?

sgorman commented:
It was implemented back in the day during the initial versions of the
project for performance enhancements as Logan mentioned the ROM does some
kind of mapping that wasn't being handled correctly, there was a big demand
by some communities for some of the tweaks to appear in the emulator
system. I think the solution is to implement the correct exception handlers
for the TLB, otherwise you will probably notice a big performance hit as
mentioned in the comments. Good luck.

- ODS

On Thu, Nov 23, 2017 at 11:35 AM, Logan <notifications@github.com> wrote:

> So potentially we could leave the hack in place for the new dynarec, since
> that is mostly targeted at mobile devices? thoughts @bsmiles32
> <https://github.com/bsmiles32> @Gillou68310
> <https://github.com/gillou68310> ?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mupen64plus/mupen64plus-core/pull/487#issuecomment-346661610>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAke39JDMwFx0Au8JGhtl8bRXiB8z-9Cks5s5Z7igaJpZM4QoUb8>
> .
>

fzurita commented:
That does bring up the question. Could this hack be applied to other TLB games with bad performance? For example, Conker's bad Fur day?

loganmc10 commented:
Ok I modified the PR to leave the new dynarec as it is (I just renamed tlb_hacks() to tlb_speed_hacks() to make it clear that it's not needed for accuracy).

loganmc10 commented:
> Could this hack be applied to other TLB games with bad performance? For example, Conker's bad Fur day?

I assume so, it looks like Conker's is doing a very similar thing (same company so it makes sense):

It does this (it's slightly different since GoldenEye actually maps the TLB into the ROM address space, Conkers looks like it's using the RDRAM address space):
```
tlb exception !!! @ 151e8620, 2, add:15019f08
map 151e8000 : 8035afff
tlb exception !!! @ 151ebb50, 2, add:151e865c
map 151eb000 : 8035bfff
tlb exception !!! @ 150432bc, 2, add:151ebcb8
map 15042000 : 802d0fff
map 15043000 : 8033efff
tlb exception !!! @ 151ec164, 2, add:151ebd04
map 151ec000 : 8035cfff
map 151ed000 : 8035ffff
tlb exception !!! @ 150417ac, 2, add:150436b0
map 15041000 : 8035dfff
tlb exception !!! @ 1517f4d8, 2, add:15019c3c
map 1517e000 : 80328fff
map 1517f000 : 8034bfff
tlb exception !!! @ 150347e8, 2, add:15018c6c
```

You can see a TLB exception happens, and then it maps the area into the TLB. If we pre-mapped those entries, the TLB exceptions would never happen.

If you look at the very top example, what you want to have happen is that when the game asks for address 0x151e8620, you want the function to return 0x8035a620 as the "TLB match"

For GoldenEye they figured out a simple pattern that would match all cases. I don't immediately see a simple pattern, but I'm not great at that sort of thing

fzurita commented:
It may explain why N64oid had so much better performance in some TLB games.

In wonder if a hack like this could be automated.

fzurita commented:
Just a thought. Is this the performance improvement that bsmiles32 fastmem is supposed to give?

fzurita commented:
I did some more research on fastmem, it seems like it would be great for TLB games. So the answer to my question would be a yes.

psyke83 commented:
Thanks - PR sent. https://github.com/mupen64plus/mupen64plus-core/pull/491
